modules/database/src/ioc/db/dbEvent.c

```diff
@@ -592,23 +592,32 @@ void db_cancel_event (dbEventSubscription event)
     UNLOCKEVQUE (que);
 
     if(sync) {
-        /* wait for worker to cycle */
+        /* cycle through worker */
         struct event_user *evUser = que->evUser;
         epicsUInt32 curSeq;
+        event_waiter wait;
+        wait.wake = epicsEventCreate(epicsEventEmpty); /* may fail */
+
         epicsMutexMustLock ( evUser->lock );
+        ellAdd(&evUser->waiters, &wait.node);
         /* grab current cycle counter, then wait for it to change */
         curSeq = evUser->pflush_seq;
         do {
             epicsMutexUnlock( evUser->lock );
-            epicsEventMustWait(evUser->pflush_sem);
-            /* The complexity needed to track the # of waiters does not seem
-             * worth it for the relatively rare situation of concurrent cancel.
-             * So uncondtionally re-trigger.  This will result in one spurious
-             * wakeup for each cancellation.
-             */
-            epicsEventTrigger(evUser->pflush_sem);
+            epicsEventMustTrigger(evUser->ppendsem);
+
+            if(wait.wake) {
+                epicsEventMustWait(wait.wake);
+            } else {
+                epicsThreadSleep(0.01); /* ick. but better than cantProceed() */
```

Comment:
I know this looks odd.  I didn't want to call `epicsEventMustCreate()`, and chance a `cantProceed()`.  This seemed like the least bad option.